source/fab/steps/grab/svn.py

```diff
@@ -3,32 +3,160 @@
+        except RuntimeError:
+            return False
+        return True
 
-        GrabSvn(
-            src='https://code.metoffice.gov.uk/svn/lfric/GPL-utilities/trunk',
-            revision=36615, dst='gpl_utils')
+
+class SvnExport(GrabSvnBase):
+    """
+    Export an FCM repo folder to the project workspace.
 
     """
     def run(self, artefact_store: Dict, config):
         super().run(artefact_store, config)
 
-        if not svn:
-            raise ImportError('svn not installed, unable to continue')
+        run_command([
+            self.command, 'export', '--force',
+            *self._cli_revision_parts(),
+            self.src,
+            str(self._dst)
+        ])
+
+
+class SvnCheckout(GrabSvnBase):
+    """
+    Checkout or update an FCM repo.
+
+    .. note::
+        If the destination is a working copy, it will be updated to the given revision, **ignoring the source url**.
+        As such, the revision should be provided via the argument, not as part of the url.
+
+    """
+    def run(self, artefact_store: Dict, config):
+        super().run(artefact_store, config)
+
+        # new folder?
+        if not self._dst.exists():  # type: ignore
+            run_command([
+                self.command, 'checkout',
+                *self._cli_revision_parts(),
+                self.src, str(self._dst)
+            ])
+
+        else:
+            # working copy?
+            if self._is_working_copy(self._dst):  # type: ignore
+                # update
+                # todo: ensure the existing checkout is from self.src?
+                run_command([self.command, 'update', *self._cli_revision_parts()], cwd=self._dst)  # type: ignore
+            else:
+                # we can't deal with an existing folder that isn't a working copy
+                raise ValueError(f"destination exists but is not an fcm working copy: '{self._dst}'")
+
+
+class SvnMerge(GrabSvnBase):
+    """
+    Merge an FCM repo into a local working copy.
+
+    """
+    def run(self, artefact_store: Dict, config):
+        super().run(artefact_store, config)
+
+        if not self._dst or not self._is_working_copy(self._dst):
+            raise ValueError(f"destination is not a working copy: '{self._dst}'")
+        else:
+            # We seem to need the url and version combined for this operation.
+            # The help for fcm merge says it accepts the --revision param, like other commands,
+            # but it doesn't seem to be recognised.
+            rev_url = f'{self.src}'
+            if self.revision is not None:
+                rev_url += f'@{self.revision}'
+
+            res = run_command([self.command, 'merge', '--non-interactive', rev_url], cwd=self._dst)
 
-        r = svn.remote.RemoteClient(self.src)
-        r.export(str(self._dst), revision=self.revision, force=True)
+            # Fcm doesn't return an error code when there's a conflict, so we have to scan the output.
+            if 'Summary of conflicts:' in res:
```

Comment:
Rather than scraping strings which may change is there a more reliable approach? Can `fcm status` be used?

Reply:
I agree, I don't like the dependency on another tool's output formatting. I didn't see an obvious better way. I'll raise a ticket to record this for future attention.

Reply:
Subversion offers an `-xml` argument which produces a well defined output you can rely on. I don't think FCM does though.

Reply:
Ah, that's better and it works with fcm too, thank you.